pygbm/gradient_boosting.py

```diff
@@ -138,62 +142,110 @@ def fit(self, X, y):
-            acc_prediction_time += toc_pred - tic_pred
-
-            acc_apply_split_time += grower.total_apply_split_time
-            acc_find_split_time += grower.total_find_split_time
+
         if self.verbose:
             duration = time() - fit_start_time
-            n_leaf_nodes = sum(p.get_n_leaf_nodes() for p in self.predictors_)
-            print(f"Fit {len(self.predictors_)} trees in {duration:.3f} s, "
-                  f"({n_leaf_nodes} total leaf nodes)")
-            print('{:<32} {:.3f}s'.format('Time spent finding best splits:',
-                                          acc_find_split_time))
-            print('{:<32} {:.3f}s'.format('Time spent applying splits:',
-                                          acc_apply_split_time))
-            print('{:<32} {:.3f}s'.format('Time spent predicting:',
-                                          acc_prediction_time))
+            n_total_leaves = sum(
+                predictor.get_n_leaf_nodes()
+                for predictors_at_ith_iteration in self.predictors_
+                for predictor in predictors_at_ith_iteration)
+            n_predictors = sum(
+                len(predictors_at_ith_iteration)
+                for predictors_at_ith_iteration in self.predictors_)
+            print(f"Fit {n_predictors} trees in {duration:.3f} s, "
+                  f"({n_total_leaves} total leaves)")
+            print(f"{'Time spent finding best splits:':<32} "
+                  f"{acc_find_split_time:.3f}s")
+            print(f"{'Time spent applying splits:':<32} "
+                  f"{acc_apply_split_time:.3f}s")
+            print(f"{'Time spent predicting:':<32} "
+                  f"{acc_prediction_time:.3f}s")
         self.train_scores_ = np.asarray(self.train_scores_)
         if self.validation_split is not None:
             self.validation_scores_ = np.asarray(self.validation_scores_)
         return self
 
     def _raw_predict(self, X):
-        """Return the sum of the leaves values"""
+        """Return the sum of the leaves values over all predictors.
+
+        The returned array (raw_predictions) is a 1d array of size
+        (n_samples * K)
+        """
         # TODO: check input / check_fitted
         # TODO: make predictor behave correctly on pre-binned data
-        raw_predictions = np.zeros(X.shape[0], dtype=np.float32)
-        for predictor in self.predictors_:
-            raw_predictions += predictor.predict(X)
+        n_samples = X.shape[0]
+        raw_predictions = np.zeros(
+            shape=(n_samples * self.n_trees_per_iteration), dtype=np.float32)
+        # Should we parallelize this?
+        for predictors_of_ith_iteration in self.predictors_:
+            for k, predictor in enumerate(predictors_of_ith_iteration):
+                start, end = n_samples * k, n_samples * (k + 1)
+                raw_predictions[start:end] += predictor.predict(X)
 
         return raw_predictions
 
     def _predict_binned(self, X_binned):
```

Comment:
Maybe this should also be renamed `_raw_predict_binned`.

Reply:
Well with the code as-is yes, but I think it's incorrect ^^

It should be returning classes for classification problems because the `predict_binned` output is directly used in `scorer._score_func`

```python
            if self.validation_split is not None:
                predicted_val = self._predict_binned(X_binned_val)
                score_val = scorer._score_func(y_val, predicted_val)
```

Tests don't fail because this is not tested, that's why I opened #55.

I say we leave it as-is even if it's buggy and fix everything when we address #55?